### ISSUE TITLE
Optimistic updates with promises

### ISSUE THREAD

**Issue by Schniz** [vigetlabs/microcosm#49]
in `9.0.0-beta-3` optimistic updates came using generators.
I think we should address `reject`ing a Promise too, to cancel the now-not-so-optimistic update.
what do you think?

I don't know yet how, but I know that it has to be treated

**Comment by nhunzaker** [comment]
I believe I've got this case covered in the latest on the v9.x branch:

https://github.com/vigetlabs/microcosm/blob/v9.x/src/__tests__/dispatch-promise-test.js#L68-L75

If a promise is rejected, the entire transaction will be rejected, so optimistic updates clean themselves up. Additionally, you can know about it failing in the callback we recently added to `app.push`:

```javascript
app.push(action, [ ...params ], function(error, body) {
    if (error) {
        console.error(error)
    }
})
```

Having said that, I'm not sure we have it published to an npm package. I'd be happy to release a new beta if you'd like to experiment with it.

**Comment by nhunzaker** [comment]
Just a bit more on this. Each time `app.push` is called, a transaction represents the action that is pushed. When the transaction fails, state will be re-determined by rerunning all outstanding valid transactions. In the case of a rejected promise, the transaction would be invalidated and would no longer be present in the newly calculated state.

**Comment by Schniz** [comment]
That's great. I'd love to experiment with that.

**Comment by nhunzaker** [comment]
I just published `9.0.0-beta-5` if you want to give it a spin :)

**Comment by nhunzaker** [comment]
Also we are working on some documentation on this here: https://github.com/vigetlabs/microcosm/blob/master/docs/guides/03-actions.md

**Comment by nhunzaker** [comment]
For the sake of being attentive: does `9.0.0-beta-5` address your needs? I'm considering a public release after adding some more comprehensive examples and copy editing the docs.

**Comment by nhunzaker** [comment]
We've released `9.0.0` and I'm fairly confident that we cover this use case. I'm going to go ahead and close this, but please reach out of we've missed something.

**Comment by Schniz** [comment]
that's great. thank you
